File: frontend/src/api/activity.ts
import { API_BASE, ApiError, api } from "./client";

export const ACTIVITY_CHANGED_EVENT = "local-gallery:activity-changed";

function notifyActivityChanged() {
  if (typeof window !== "undefined") {
    window.dispatchEvent(new Event(ACTIVITY_CHANGED_EVENT));
  }
}

async function withActivityChanged<T>(
  request: Promise<T>,
  affectsUnread = true,
): Promise<T> {
  const result = await request;
  if (affectsUnread) notifyActivityChanged();
  return result;
}

export function isActivityNotFound(error: unknown): boolean {
  return (
    error instanceof ApiError &&
    error.status === 404 &&
    error.code === "activity_not_found"
  );
}

export type MediaKind = "image" | "video";
export type ActivityStatus = "in_progress" | "completed";

export interface ActivityIdentity {
  albumId: string;
  mediaKind: MediaKind;
  itemId?: string;
}

export interface Activity extends ActivityIdentity {
  pageIndex?: number;
  pageCount?: number;
  positionMs?: number;
  durationMs?: number;
  status: ActivityStatus;
  updated: string;
}

export interface ImageActivity extends Activity {
  mediaKind: "image";
  pageIndex: number;
  pageCount: number;
}

export interface VideoActivity extends Activity {
  mediaKind: "video";
  itemId: string;
  positionMs: number;
  durationMs: number;
}

export type ActivityInput =
  | {
      albumId: string;
      mediaKind: "image";
      pageIndex: number;
      pageCount: number;
    }
  | {
      albumId: string;
      mediaKind: "video";
      itemId: string;
      positionMs: number;
      durationMs: number;
    };

export const activityApi = {
  get: (identity: ActivityIdentity) =>
    api<Activity>("/api/activity", {
      params: {
        albumId: identity.albumId,
        mediaKind: identity.mediaKind,
        itemId: identity.itemId,
      },
    }),

  getImage: (albumId: string) =>
    api<ImageActivity>("/api/activity", {
      params: { albumId, mediaKind: "image" },
    }),

  getVideo: (albumId: string, itemId: string) =>
    api<VideoActivity>("/api/activity", {
      params: { albumId, mediaKind: "video", itemId },
    }),

  set: (activity: ActivityInput) =>
    withActivityChanged(
      api<Activity>("/api/activity", { method: "PUT", body: activity }),
      activity.mediaKind === "image",
    ),

  setImage: (albumId: string, pageIndex: number, pageCount: number) =>
    withActivityChanged(
      api<ImageActivity>("/api/activity", {
        method: "PUT",
        body: { albumId, mediaKind: "image", pageIndex, pageCount },
      }),
    ),

  setVideo: (
    albumId: string,
    itemId: string,
    positionMs: number,
    durationMs: number,
  ) =>
    api<VideoActivity>("/api/activity", {
      method: "PUT",
      body: {
        albumId,
        mediaKind: "video",
        itemId,
        positionMs,
        durationMs,
      },
    }),

  query: (items: ActivityIdentity[]) =>
    api<{ activities: Activity[]; count: number }>("/api/activity/query", {
      method: "POST",
      body: { items },
    }),

  queryImages: async (albumIds: string[]) => {
    const response = await api<{ activities: ImageActivity[]; count: number }>(
      "/api/activity/query",
      {
        method: "POST",
        body: {
          items: albumIds.map((albumId) => ({
            albumId,
            mediaKind: "image" as const,
          })),
        },
      },
    );
    return response;
  },

  setBatch: (activities: ActivityInput[]) =>
    withActivityChanged(
      api<{ ok: boolean; updated: number }>("/api/activity/batch", {
        method: "PUT",
        body: { activities },
      }),
      activities.some((activity) => activity.mediaKind === "image"),
    ),

  remove: (identity: ActivityIdentity) =>
    withActivityChanged(
      api<{ ok: boolean; removed: boolean }>("/api/activity", {
        method: "DELETE",
        params: {
          albumId: identity.albumId,
          mediaKind: identity.mediaKind,
          itemId: identity.itemId,
        },
      }),
      identity.mediaKind === "image",
    ),

  removeImage: (albumId: string) =>
    withActivityChanged(
      api<{ ok: boolean; removed: boolean }>("/api/activity", {
        method: "DELETE",
        params: { albumId, mediaKind: "image" },
      }),
    ),

  clear: () =>
    withActivityChanged(
      api<{ ok: boolean; removed: number }>("/api/activity/all", {
        method: "DELETE",
      }),
    ),

  // 页面卸载时的最后一次写入不等待响应；会话仍只通过 HttpOnly cookie 携带。
  keepalive: (activity: ActivityInput) => {
    void fetch(`${API_BASE}/api/activity`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
      },
      credentials: "include",
      keepalive: true,
      body: JSON.stringify(activity),
    }).then((response) => {
      if (response.ok && activity.mediaKind === "image") {
        notifyActivityChanged();
      }
    }).catch(() => {});
  },
};

File: backend/internal/models/activity.go
package models

import (
	"errors"
	"strings"
	"time"
)

// MediaKind 区分活动记录的媒体语义。图片保存相册页码，视频保存单个文件的播放时间。
type MediaKind string

const (
	MediaKindImage MediaKind = "image"
	MediaKindVideo MediaKind = "video"
)

// ActivityStatus 是服务端根据位置派生的状态，客户端不能单独篡改。
type ActivityStatus string

const (
	ActivityInProgress ActivityStatus = "in_progress"
	ActivityCompleted  ActivityStatus = "completed"
)

// Activity 统一表示图片阅读与视频播放活动。
//
// 图片活动的唯一键是 albumId + mediaKind；视频活动再加 itemId，
// 因此同一混合相册的图片页码与每个视频的播放位置互不覆盖。
type Activity struct {
	AlbumID    string         `json:"albumId"`
	MediaKind  MediaKind      `json:"mediaKind"`
	ItemID     string         `json:"itemId,omitempty"`
	PageIndex  int            `json:"pageIndex"`
	PageCount  int            `json:"pageCount"`
	PositionMS int64          `json:"positionMs"`
	DurationMS int64          `json:"durationMs"`
	Status     ActivityStatus `json:"status"`
	Updated    time.Time      `json:"updated"`
}

// ActivityIdentity 定位一条活动，不包含可变的进度值。
type ActivityIdentity struct {
	AlbumID   string    `json:"albumId"`
	MediaKind MediaKind `json:"mediaKind"`
	ItemID    string    `json:"itemId,omitempty"`
}

// IsFileID 校验对外媒体文件 ID。
func IsFileID(value string) bool {
	return isOpaqueResourceID(value, "f_")
}

// NormalizeActivity 校验互斥字段并由实际位置派生完成状态。
func NormalizeActivity(activity Activity) (Activity, error) {
	if !IsAlbumID(activity.AlbumID) {
		return Activity{}, errors.New("invalid activity album id")
	}
	switch activity.MediaKind {
	case MediaKindImage:
		if activity.ItemID != "" || activity.PageCount <= 0 || activity.PageIndex < 0 ||
			activity.PageIndex >= activity.PageCount || activity.PositionMS != 0 || activity.DurationMS != 0 {
			return Activity{}, errors.New("invalid image activity")
		}
		if activity.PageIndex == activity.PageCount-1 {
			activity.Status = ActivityCompleted
		} else {
			activity.Status = ActivityInProgress
		}
	case MediaKindVideo:
		if !IsFileID(activity.ItemID) || activity.PageIndex != 0 || activity.PageCount != 0 ||
			activity.PositionMS < 0 || activity.DurationMS <= 0 || activity.PositionMS > activity.DurationMS {
			return Activity{}, errors.New("invalid video activity")
		}
		// 播放到 98% 视为完成，避免片尾黑帧或编码时长误差让记录永远停在未完成。
		if activity.PositionMS > 0 && activity.PositionMS >= activity.DurationMS-(activity.DurationMS/50) {
			activity.Status = ActivityCompleted
		} else {
			activity.Status = ActivityInProgress
		}
	default:
		return Activity{}, errors.New("invalid activity media kind")
	}
	return activity, nil
}

// Key 返回仅用于持久化索引的稳定键。键中只有不透明 ID，不会泄露路径。
func (identity ActivityIdentity) Key() (string, bool) {
	if !IsAlbumID(identity.AlbumID) {
		return "", false
	}
	switch identity.MediaKind {
	case MediaKindImage:
		if identity.ItemID != "" {
			return "", false
		}
		return string(identity.MediaKind) + ":" + identity.AlbumID, true
	case MediaKindVideo:
		if !IsFileID(identity.ItemID) {
			return "", false
		}
		return strings.Join([]string{string(identity.MediaKind), identity.AlbumID, identity.ItemID}, ":"), true
	default:
		return "", false
	}
}

// Identity 取出活动的不变定位字段。
func (activity Activity) Identity() ActivityIdentity {
	return ActivityIdentity{
		AlbumID:   activity.AlbumID,
		MediaKind: activity.MediaKind,
		ItemID:    activity.ItemID,
	}
}
